File: aws/s3/typescript/src/services/presign.ts
/**
 * Presign Service
 *
 * Generate presigned URLs for S3 operations.
 */

import { S3Config, resolveEndpoint, buildPath } from "../config";
import { AwsSignerV4 } from "../signing";
import { PresignGetRequest, PresignPutRequest, PresignDeleteRequest } from "../types/requests";
import { PresignedUrl } from "../types/responses";

/**
 * Presign service for generating presigned URLs.
 */
export class PresignService {
  private config: S3Config;
  private signer: AwsSignerV4;

  constructor(config: S3Config, signer: AwsSignerV4) {
    this.config = config;
    this.signer = signer;
  }

  /**
   * Generate a presigned GET URL.
   */
  async presignGet(request: PresignGetRequest): Promise<PresignedUrl> {
    return this.presign("GET", request.bucket, request.key, request.expiresIn, {
      versionId: request.versionId,
      responseContentType: request.responseContentType,
      responseContentDisposition: request.responseContentDisposition,
    });
  }

  /**
   * Generate a presigned PUT URL.
   */
  async presignPut(request: PresignPutRequest): Promise<PresignedUrl> {
    const headers: Record<string, string> = {};

    if (request.contentType) {
      headers["content-type"] = request.contentType;
    }
    if (request.storageClass) {
      headers["x-amz-storage-class"] = request.storageClass;
    }

    return this.presign("PUT", request.bucket, request.key, request.expiresIn, {}, headers);
  }

  /**
   * Generate a presigned DELETE URL.
   */
  async presignDelete(request: PresignDeleteRequest): Promise<PresignedUrl> {
    return this.presign("DELETE", request.bucket, request.key, request.expiresIn, {
      versionId: request.versionId,
    });
  }

  /**
   * Generate a presigned HEAD URL.
   */
  async presignHead(bucket: string, key: string, expiresIn: number): Promise<PresignedUrl> {
    return this.presign("HEAD", bucket, key, expiresIn);
  }

  private async presign(
    method: string,
    bucket: string,
    key: string,
    expiresIn: number,
    queryParams?: Record<string, string | undefined>,
    headers?: Record<string, string>
  ): Promise<PresignedUrl> {
    // Validate expiration (max 7 days)
    const maxExpiration = 7 * 24 * 60 * 60;
    if (expiresIn > maxExpiration) {
      expiresIn = maxExpiration;
    }

    const endpoint = resolveEndpoint(this.config, bucket);
    const path = buildPath(this.config, bucket, key);
    const url = new URL(path, endpoint);

    // Add query parameters
    if (queryParams) {
      for (const [key, value] of Object.entries(queryParams)) {
        if (value !== undefined) {
          if (key === "responseContentType") {
            url.searchParams.set("response-content-type", value);
          } else if (key === "responseContentDisposition") {
            url.searchParams.set("response-content-disposition", value);
          } else if (key === "versionId") {
            url.searchParams.set("versionId", value);
          }
        }
      }
    }

    const presignedUrl = await this.signer.presign(method, url, headers ?? {}, expiresIn);

    const expiresAt = new Date(Date.now() + expiresIn * 1000);

    return {
      url: presignedUrl.toString(),
      method,
      expiresAt,
      signedHeaders: headers ?? {},
    };
  }
}

File: aws/s3/rust/src/services/presign.rs
//! Presign service for generating presigned URLs.

use crate::config::S3Config;
use crate::error::S3Error;
use crate::signing::AwsSigner;
use crate::types::*;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;
use url::Url;

/// Service for generating presigned URLs.
pub struct PresignService {
    config: Arc<S3Config>,
    signer: Arc<dyn AwsSigner>,
}

impl PresignService {
    /// Create a new presign service.
    pub fn new(config: Arc<S3Config>, signer: Arc<dyn AwsSigner>) -> Self {
        Self { config, signer }
    }

    /// Generate a presigned URL for GET (download).
    pub async fn presign_get(
        &self,
        request: PresignGetRequest,
    ) -> Result<PresignedUrl, S3Error> {
        let mut query_params = Vec::new();

        if let Some(version_id) = &request.version_id {
            query_params.push(format!("versionId={}", version_id));
        }
        if let Some(content_type) = &request.response_content_type {
            query_params.push(format!("response-content-type={}", content_type));
        }
        if let Some(disposition) = &request.response_content_disposition {
            query_params.push(format!("response-content-disposition={}", disposition));
        }

        let query = if query_params.is_empty() {
            None
        } else {
            Some(query_params.join("&"))
        };

        let url = self.build_url(&request.bucket, Some(&request.key), query.as_deref())?;

        self.signer.presign("GET", &url, request.expires_in, None).await
    }

    /// Generate a presigned URL for PUT (upload).
    pub async fn presign_put(
        &self,
        request: PresignPutRequest,
    ) -> Result<PresignedUrl, S3Error> {
        let url = self.build_url(&request.bucket, Some(&request.key), None)?;

        let mut headers = HashMap::new();

        if let Some(content_type) = &request.content_type {
            headers.insert("content-type".to_string(), content_type.clone());
        }
        if let Some(content_length) = request.content_length {
            headers.insert("content-length".to_string(), content_length.to_string());
        }
        if let Some(storage_class) = &request.storage_class {
            headers.insert(
                "x-amz-storage-class".to_string(),
                storage_class.as_str().to_string(),
            );
        }

        let additional_headers = if headers.is_empty() {
            None
        } else {
            Some(&headers)
        };

        self.signer
            .presign("PUT", &url, request.expires_in, additional_headers)
            .await
    }

    /// Generate a presigned URL for DELETE.
    pub async fn presign_delete(
        &self,
        request: PresignDeleteRequest,
    ) -> Result<PresignedUrl, S3Error> {
        let mut query_params = Vec::new();

        if let Some(version_id) = &request.version_id {
            query_params.push(format!("versionId={}", version_id));
        }

        let query = if query_params.is_empty() {
            None
        } else {
            Some(query_params.join("&"))
        };

        let url = self.build_url(&request.bucket, Some(&request.key), query.as_deref())?;

        self.signer.presign("DELETE", &url, request.expires_in, None).await
    }

    /// Generate a presigned URL for HEAD (metadata).
    pub async fn presign_head(
        &self,
        bucket: &str,
        key: &str,
        expires_in: Duration,
    ) -> Result<PresignedUrl, S3Error> {
        let url = self.build_url(bucket, Some(key), None)?;
        self.signer.presign("HEAD", &url, expires_in, None).await
    }

    fn build_url(
        &self,
        bucket: &str,
        key: Option<&str>,
        query: Option<&str>,
    ) -> Result<Url, S3Error> {
        let endpoint = self.config.resolve_endpoint(Some(bucket));
        let path = self.config.build_path(bucket, key);

        let url_str = if let Some(q) = query {
            format!("{}{}?{}", endpoint.as_str().trim_end_matches('/'), path, q)
        } else {
            format!("{}{}", endpoint.as_str().trim_end_matches('/'), path)
        };

        Url::parse(&url_str).map_err(|e| {
            S3Error::Request(crate::error::RequestError::Validation {
                message: format!("Invalid URL: {}", e),
            })
        })
    }
}

impl std::fmt::Debug for PresignService {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PresignService")
            .field("config", &self.config)
            .finish_non_exhaustive()
    }
}
